### PR TITLE
Remove deprecated datamodule lifecycle properties

### LINKED ISSUE #10312: Remove deprecated code after the 1.5 release
## Proposed refactoring or deprecation

Remove deprecated code after the 1.5 release

### Pitch

All the deprecated features we have are tested here

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/deprecated_api/test_remove_1-6.py

If you are interested in taking care of one item, post a comment here asking to take it. This avoids multiple people working on the same thing.

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by mayeroa**
## What does this PR do?

Part of #10312

### Does your PR introduce any breaking changes? If yes, please list them.

They are deprecated

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]

**Review by ananthsub — CHANGES_REQUESTED** [review]
Thanks for working on this @mayeroa !
With this change, we should also remove the following:
- https://github.com/PyTorchLightning/pytorch-lightning/blob/ce3e63262a7247c1111375b28c90bdd83b42bf3f/pytorch_lightning/core/datamodule.py#L94-L105
- https://github.com/PyTorchLightning/pytorch-lightning/blob/ce3e63262a7247c1111375b28c90bdd83b42bf3f/pytorch_lightning/core/datamodule.py#L409-L483
- Any references to these properties in the trainer, like here: https://github.com/PyTorchLightning/pytorch-lightning/blob/ce3e63262a7247c1111375b28c90bdd83b42bf3f/pytorch_lightning/trainer/connectors/data_connector.py#L143

File: tests/core/test_datamodules.py [hunk 1]
```diff
@@ -140,110 +140,110 @@ def test_helper_boringdatamodule_with_verbose_setup():
...
-    assert not dm.has_prepared_data
-    assert not dm.has_setup_fit
-    assert not dm.has_setup_test
-    assert not dm.has_setup_validate
-    assert not dm.has_setup_predict
-    assert not dm.has_teardown_fit
-    assert not dm.has_teardown_test
-    assert not dm.has_teardown_validate
-    assert not dm.has_teardown_predict
+    assert not dm._has_prepared_data
```
Comment (ananthsub) [inline tests/core/test_datamodules.py:143]: we should be able to remove this whole test now

File: tests/core/test_datamodules.py [hunk 2]
```diff
@@ -140,110 +140,110 @@ def test_helper_boringdatamodule_with_verbose_setup():
...
     dm.prepare_data()
-    assert not dm.has_setup_fit
-    assert not dm.has_setup_test
-    assert not dm.has_setup_validate
-    assert not dm.has_setup_predict
-    assert not dm.has_teardown_fit
-    assert not dm.has_teardown_test
-    assert not dm.has_teardown_validate
-    assert not dm.has_teardown_predict
+    assert not dm._has_setup_fit
```
Comment (ananthsub) [inline tests/core/test_datamodules.py:191]: same for this test

**Comment by mayeroa** [comment]
Should be better now :)

File: docs/source/extensions/datamodules.rst [hunk 3]
```diff
@@ -237,8 +237,7 @@ we assume all stages have been set-up.
 .. note::
     ``{setup,teardown,prepare_data}`` call will be only called once for a specific stage.
     If the stage was ``None`` then we assume ``{fit,validate,test}`` have been called. For example, this means that
-    any duplicate ``dm.setup('fit')`` calls will be a no-op. To avoid this, you can overwrite
-    ``dm._has_setup_fit = False``
+    any duplicate ``dm.setup('fit')`` calls will be a no-op.
```
Comment (ananthsub) [inline docs/source/extensions/datamodules.rst]: This note must be removed now as well as this removal changes this behavior

```
note::
    ``{setup,teardown,prepare_data}`` call will be only called once for a specific stage.
    If the stage was ``None`` then we assume ``{fit,validate,test}`` have been called. For example, this means that
    any duplicate ``dm.setup('fit')`` calls will be a no-op. 
```

**Review by carmocca — APPROVED** [review]
Big one!

**Review by ananthsub — APPROVED** [review]
